typing/typemod.ml

```diff
@@ -2427,58 +2439,93 @@ and type_structure ?(toplevel = false) funct_body anchor env sstr =
         in
         (* Note: Env.find_value does not trigger the value_used event. Values
            will be marked as being used during the signature inclusion test. *)
+        let items, shape_map =
+          List.fold_left
+            (fun (acc, shape_map) (id, { Asttypes.loc; _ }, _typ)->
+              Signature_names.check_value names loc id;
+              let vd =  Env.find_value (Pident id) newenv in
+              Env.register_uid vd.val_uid vd.val_loc;
+              Sig_value(id, vd, Exported) :: acc,
+              Shape.Map.add_value shape_map id vd.val_uid
+            )
+            ([], shape_map)
+            (let_bound_idents_full defs)
+        in
         Tstr_value(rec_flag, defs),
-        List.map (fun (id, { Asttypes.loc; _ }, _typ)->
-          Signature_names.check_value names loc id;
-          Sig_value(id, Env.find_value (Pident id) newenv, Exported)
-        ) (let_bound_idents_full defs),
+        List.rev items,
+        shape_map,
         newenv
     | Pstr_primitive sdesc ->
         let (desc, newenv) = Typedecl.transl_value_decl env loc sdesc in
         Signature_names.check_value names desc.val_loc desc.val_id;
+        Env.register_uid desc.val_val.val_uid desc.val_val.val_loc;
         Tstr_primitive desc,
         [Sig_value(desc.val_id, desc.val_val, Exported)],
+        Shape.Map.add_value shape_map desc.val_id desc.val_val.val_uid,
         newenv
     | Pstr_type (rec_flag, sdecls) ->
         let (decls, newenv) = Typedecl.transl_type_decl env rec_flag sdecls in
         List.iter
           Signature_names.(fun td -> check_type names td.typ_loc td.typ_id)
           decls;
-        Tstr_type (rec_flag, decls),
-        map_rec_type_with_row_types ~rec_flag
+        let items = map_rec_type_with_row_types ~rec_flag
           (fun rs info -> Sig_type(info.typ_id, info.typ_type, rs, Exported))
-          decls [],
+          decls []
+        in
+        let shape_map = List.fold_left
+          (fun shape_map -> function
+            | Sig_type (id, vd, _, _) ->
+              Env.register_uid vd.type_uid vd.type_loc;
+              Shape.Map.add_type shape_map id vd.type_uid
+            | _ -> assert false
+          )
+          shape_map
+          items
```

Comment:
Fair point, I had missed the fact that regular type declarations generated more items than present in the source!

Actually, I think our handling of classes is not quite right: we should register the various items in our shapes, since they can appear in the source.
Which raises the question: can the `...#row` types generated here be adressed/referred in the source? That will decide whether we want to store them in the shape map or not; but for now I haven't been able to answer (my guess is: no).

Reply:
I don't think it is possible to obtain a syntactic item that points to a private `...#row` type, at the very least this is not intended to be possible.

Reply:
Ok, I have done that in 
fd1878b